core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1155 @@
+/// for safe addition.
+impl ops::Add for Fixed64 {
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self::Output {
+		Self(self.0 + rhs.0)
+	}
+}
+
+/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
+/// for safe subtraction.
+impl ops::Sub for Fixed64 {
+	type Output = Self;
+
+	fn sub(self, rhs: Self) -> Self::Output {
+		Self(self.0 - rhs.0)
+	}
+}
+
+impl CheckedSub for Fixed64 {
+	fn checked_sub(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_sub(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+impl CheckedAdd for Fixed64 {
+	fn checked_add(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_add(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+/// Helper gcd function used in Rational128 implementation.
+fn gcd(a: u128, b: u128) -> u128 {
+	match ((a, b), (a & 1, b & 1)) {
+		((x, y), _) if x == y => y,
+		((0, x), _) | ((x, 0), _) => x,
+		((x, y), (0, 1)) | ((y, x), (1, 0)) => gcd(x >> 1, y),
+		((x, y), (0, 0)) => gcd(x >> 1, y >> 1) << 1,
+		((x, y), (1, 1)) => {
+			let (x, y) = (min(x, y), max(x, y));
+			gcd((y - x) >> 1, x)
+		},
+		_ => unreachable!(),
+    }
+}
+
+/// A wrapper for any rational number with a 128 bit numerator and denominator.
+#[derive(Clone, Copy, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Rational128(u128, u128);
+
+/// Some helper functions to work with 128bit numbers. Note that the functionality provided here is
+/// only sensible to use with 128bit numbers because for smaller sizes, you can always rely on
+/// assumptions of a bigger type (u128) being available, or simply create a per-thing and use the
+/// multiplication implementation provided there.
+pub mod helpers_128bit {
+	use super::Perquintill;
+	use crate::traits::Zero;
+
+	/// return the number of bits that are needed to store `n`.
+	pub(crate) fn bits_of(n: u128) -> u32 {
+		128 - n.leading_zeros()
+	}
+
+	/// Tries to compute `a * b / c`. The approach is:
+	///   - Simply try `a * b / c`.
+	///   - Else: Swap the operations (divide first) if `a > c` (division is possible) and `b <= c`
+	///     (overflow cannot happen)
+	///   - Else: Try and compute a perquintillion from `b / c` if `b <= c`.
+	///   - Else: return `Err`.
+	///
+	/// If none worked then return Error.
+	///
+	/// c must be greater than or equal to 1.
+	pub fn multiply_by_rational(a: u128, b: u128, c: u128) -> Result<u128, &'static str> {
+		if a.is_zero() { return Ok(Zero::zero()); }
+		let c = c.max(1);
+
+		if let Some(x) = a.checked_mul(b) {
+			// This is the safest way to go. Try it.
+			Ok(x / c)
+		} else if a > c && b <= c {
+			// if it can be safely swapped and it is a fraction, then swap. The operations in this
+			// block cannot overflow since b <= c.
+			let q = a / c;
+			let r = a % c;
+			let r_additional = (r * b) / c; // todo check overflow
+			Ok(q * b + r_additional) // todo check overflow
```

Comment:
TODO ?

Reply:
no I just forgot to remove them. The comment above defines why it cannot overflow. Nonetheless, I've made the non-obvious ones saturating. the `(q * b)` is for sure safe because `q = a / c` and `b < c`.

This is actually the main difference now to compensate for the accuracy loss of this swapping method. Should be interesting for you to also see the test cases for this. The edge-case you reported earlier is now covered. (like `(multiply_by_rational(2*MAX64 - 1, MAX64, MAX64).unwrap()`)